### DISCUSSION TITLE
Improve ergonomics of SwitchStore with no action

### DISCUSSION THREAD

**Discussion by AndreiVidrasco** [pointfreeco/swift-composable-architecture#660] [General]
**Describe the bug**
Hi. Is it possible to somehow ergonomics when SwitchStore doesn't need any actions?
See example below. Basically, I want to have a way to get rid of all this action closure. 
At the moment it doesn't compile, if I remove. 

```swift
struct ContentViewState: Equatable {
    let viewTypes: DifferentViewTypes
}

enum DifferentViewTypes: Equatable {
    case noData
    case genericError
    case serverResponse(String)
}

struct ContentView: View {
    var store: Store<ContentViewState, Never>
    
    var body: some View {
        SwitchStore(store.scope(state: \.viewTypes, action: { i in fatalError() })) {
            CaseLet(state: /DifferentViewTypes.genericError, action: { _ in }, then: GenericErrorView.init)
            CaseLet(state: /DifferentViewTypes.noData, action: { _ in }, then: NoDataView.init)
            CaseLet(state: /DifferentViewTypes.serverResponse, action: { _ in }, then: ServerResponseView.init)
        }
    }
}

struct NoDataView: View {
    var store: Store<(), Never>
    
    var body: some View {
        Text("No Data")
    }
}

struct GenericErrorView: View {
    var store: Store<(), Never>
    
    var body: some View {
        Text("Generic Error")
    }
}


struct ServerResponseView: View {
    var store: Store<String, Never>
    
    var body: some View {
        WithViewStore(store) { viewStore in
            Text(viewStore.capitalized)
        }
    }
}
```

**Expected behavior**
Have a way to not include actions conversion in case of never

**Comment by mbrandonw** [comment]
I am going to convert this to a discussion as it is not an issue with the library.

**Comment by mbrandonw** [comment]
You can perform `self.store.actionless` to turn any store into one that has `Never` for its actions:

https://github.com/pointfreeco/swift-composable-architecture/blob/e7dda73c35f1016c8ba82fd2b7c43757cce68e58/Sources/ComposableArchitecture/Store.swift#L407

Also, the `action` argument of `.scope` is optional, so you can leave that off once you apply `.actionless`.

**Reply by iampatbrown** [reply]
I'm having a look now. Might have to add an extra init on CaseLet View where Action == Never
self.store.actionless didn't work when I tried it

**Reply by iampatbrown** [reply]
I'm going to check this out tomorrow. I thought actionless would have worked...

**Reply by AndreiVidrasco** [reply]
It's already actionless. I know that parameter 'action' is optional, however it doesn't compile if I remove it.

**Reply by iampatbrown** [reply]
This seems to work for me.
```swift
public extension CaseLet where GlobalAction == Never, LocalAction == Never {
  init(
    state toLocalState: @escaping (GlobalState) -> LocalState?,
    @ViewBuilder then content: @escaping (Store<LocalState, LocalAction>) -> Content
  ) {
    func absurd<A>(_ never: Never) -> A {}
    self.init(state: toLocalState, action: absurd, then: content)
  }
}
```

```swift
struct ContentView: View {
  var store: Store<ContentViewState, Never>

      var body: some View {
        SwitchStore(store.scope(state: \.viewTypes)) {
              CaseLet(state: /DifferentViewTypes.genericError, then: GenericErrorView.init)
              CaseLet(state: /DifferentViewTypes.noData, then: NoDataView.init)
              CaseLet(state: /DifferentViewTypes.serverResponse, then: ServerResponseView.init)
          }
      }
}
```

**Reply by AndreiVidrasco** [reply]
ok, that looks like it. is it possible to add in the main library?
also just a suggestion, maybe to add also this extension
```swift
extension CaseLet {
    public init(
      state toLocalState: @escaping (GlobalState) -> LocalState?,
      @ViewBuilder then content: @escaping () -> Content
    ) where LocalAction == Never, GlobalAction == Never, LocalState == Void {
        func absurd<A>(_ never: Never) -> A {}
        self.init(state: toLocalState, action: absurd, then: { _ in content() })
    }
}
```

in case there is no action and no state, we can remove store from content block.
then I can remove `var store: Store<(), Never>` inside the View and also keep the nice syntax of `then: GenericErrorView.init)`

**Reply by mbrandonw** [reply]
Sorry about that! Sent my comment from my phone and so I didn't catch the nuances of the problem. Y'all are both right that an overload of `CaseLet.init` is needed, so we'll discuss adding that to the library today!

**Reply by iampatbrown** [reply]
Sounds good :)

**Reply by mbrandonw** [reply]
After discussing we decided to not add this overload. Extra overloads come with autocomplete fatigue, and this one in particular seems to only service a very small set of situations. We have other overloads in the library that we'd also like to get rid of, and may soon start deprecating them.

FWIW, you can slightly shorten your code to this:

```swift
SwitchStore(store.scope(state: \.viewTypes)) {
  CaseLet(state: /DifferentViewTypes.genericError, action: { _ in }, then: GenericErrorView.init)
  CaseLet(state: /DifferentViewTypes.noData, action: { _ in }, then: NoDataView.init)
  CaseLet(state: /DifferentViewTypes.serverResponse, action: { _ in }, then: ServerResponseView.init)
}
```

And you can shorten more if you bring in the convenience initializer into your own code base.
